### DISCUSSION TITLE
Discussion: target-typed slicing

### DISCUSSION THREAD

**Discussion by alrz** [dotnet/csharplang#6390] [Language Ideas]
Currently, all slicing APIs return the same type as the container's.
```cs
int[] slice = array[..];
```
Even if you specify the target-type, the allocation will happen under the hood,
```cs
Span<int> slice = array[..]; // allocates another array
```
Slice patterns suffer from the same problem:
```cs
array is [1, .. var slice] // allocates another array
```
The workaround is to remember to always use `AsSpan` before doing any slicing operations.

--

While binding an *element_access* like this:
```cs
Span<int> a = sequence[..];
```
The compiler will lookup the following members:

1. `this[Range]` (instance-only)
2. `Slice(int, int)` (instance-only)
3. `AsSpan(int, int)` (instance or extension)

The `AsSpan` method is preferred only if there is a target-type and it's identical to the return-type of the method.

> **Open question**: Should an implicit conversion also work?

Same rules can be applied to slice patterns, for example:

```cs
sequence is [.. Span<int> span]
```
The compiler will assign `sequence.AsSpan(0, length)` to the variable `span` if there's no `Slice` or it cannot be used.

Note: This is an existing API as `AsSpan` is already implemented for `T[]`, `ImmutableArray<T>`, etc but it may not apply to all types with slicing support.

**Comment by svick** [comment]
To me, this seems like it would be too much of a breaking change. E.g., if I understand it correctly, it would change the output of the following code:

```c#
int[] array = new int[1];
Span<int> slice = array[..];
array[0] = 42;
Console.WriteLine(slice[0]);
```

**Reply by alrz** [reply]
Definitely a breaking change, but to be honest that's the behavior I would (and did) expect from this code, as with other languages. 

```cs
Span<int> slice = array[..];
````
In my opinion, the fact that an implicit copy is happening here is completely against what the developer intended.

**Comment by HaloFour** [comment]
@alrz 

I recall this behavior being described as intentional and desirable as slicing is expected to follow a given pattern.  If you want a `Span<T>` you can use `array.AsSpan<T>()[..]`.

**Reply by alrz** [reply]
Runtime is consistent in how slicing is declared. But I believe that implicit copy is an unintended consequence of a conversion which is meant to be used for assigning arrays to a span: `Span<int> s = new [] {};`, not ~~incurring~~ hiding a copy.

**Reply by HaloFour** [reply]
I guess the big question is how far the cows already got out of the barn.  For the case of inferring the target type I doubt there's much that can be done, but if the target type is already `Span<T>` I can see where it makes sense to try to narrowly support a different pattern to avoid that implicit copy.  However, that could also lead to subtle bugs where the copy semantics are currently relied on.  Even that feels very risky.

**Reply by alrz** [reply]
I would love to see some data on this. A [text search](https://sourcegraph.com/search?q=context:global+repo:%5Egithub%5C.com/dotnet+%28Span%5C%3C%5Cw%2B%5C%3E+%5Cw%2B+%5C%3D+%5Cw%2B%5C%5B%5B%5E%5C.%5C%5D%5D*%5C.%7B2%7D%29&patternType=regexp) didn't yield any useful results.


But I did find that there's an analyzer for this: [CA1833](https://docs.microsoft.com/en-us/dotnet/fundamentals/code-analysis/quality-rules/ca1833) so the impact is already limited.

**Reply by alrz** [reply]
Which, doesn't make my argument any stronger. Maybe this could only apply to slice patterns without adding target-typing to the mix.

**Comment by TahirAhmadov** [comment]
We definitely need a list of items to "do perfectly" when we get around to re-writing C# from scratch. :)

**Reply by alrz** [reply]
I don't think that would change anything here. It's just that implicit conversions can be too broad but the convenience is seemingly preferred over a surprising behavior in a few edge cases. We have analyzers, so generally this wouldn't be too much of a concern.

**Comment by yaakov-h** [comment]
> The `AsSpan` method is preferred only if there is a target-type and it's identical to the return-type of the method.

So `Span` would be non-allocating but `ReadOnlySpan` would still take a copy?

**Reply by alrz** [reply]
`Open question: Should an implicit conversion also work?`

Even if the breaking change is acceptable, I'd be wary to permit all implicit conversions. Maybe this could only work with any span/memory type, but nothing more.
